### ISSUE TITLE
JSON API inconsistencies

### ISSUE THREAD

**Issue by ewdurbin** [pypa/warehouse#3206]
re https://github.com/pypa/pipenv/issues/1646#issuecomment-371972522

```
ewdurbin$ curl -s https://pypi.python.org/pypi/django-oidc-provider/json | jq '.info.requires_dist'
null
ewdurbin$ curl -s https://pypi.org/pypi/django-oidc-provider/json | jq '.info.requires_dist'
[]
```

**Comment by kennethreitz** [comment]
I'm stealing the "emoji in GH issues" idea from here

**Comment by ewdurbin** [comment]
Doing some quick analysis to try to capture everything we can in one issue and squash them all at once.

**Comment by ewdurbin** [comment]
Key syntax from [`jq`](https://stedolan.github.io/jq/)

Missing in warehouse implementation (probably not an issue):
  - `.info._pypi_hidden`
  - `.info._pypi_ordering`
  - `.info.cheesecake_code_kwalitee_id`
  - `.info.cheesecake_documentation_id`
  - `.info.cheesecake_installability_id`

Renamed in warehouse implementation:
  - `.info.package_url` => `.info.project_url`

Added in warehouse implementation:
  - `.releases[][].digests.md5`
  - `.releases[][].digests.sha256`

Behavior change in warehouse implementation:
  - if a project/release has no `requires_dist` values, an empty list is returned. this is a change in that previously no key named `requires_dist` was set.

**Comment by dstufft** [comment]
> Missing in warehouse implementation (probably not an issue):
>   - `.info._pypi_hidden`
>   - `.info._pypi_ordering`
>   - `.info.cheesecake_code_kwalitee_id`
>   - `.info.cheesecake_documentation_id`
>   - `.info.cheesecake_installability_id`

These were purposely left out, and shouldn't be re-added unless we have evidence of wide-spread breakage without them. They're either internal details or removed features.

> Renamed in warehouse implementation:
>   - `.info.package_url` => `.info.project_url`

I don't remember why this was changed, if there was a reason at all.  Probably this should be changed back (possibly keep both spellings of the name in case someone was using the Warehouse version).


> Added in warehouse implementation:
>  - `.releases[][].path`

I don't see a path key? It shouldn't be exposed in either case, it's an implementation detail.

>  - `.releases[][].digests.md5`
>  - `.releases[][].digests.sha256`

These are fine (and should be the preferred spelling) so that we provide a path forward for sha256 hashes in the API (and others).

> Behavior change in warehouse implementation:
>  - if a project/release has no `.info.requires_dist` values, an empty list is returned. this is a change in that previously no key named `info.requires_dist` existed.

I think the Warehouse behavior makes a lot more sense, and I highly suspect very little software out there is relying on the difference (and really, you shouldn't be relying on a key *not* existing in JSON output).

**Comment by ewdurbin** [comment]
These are fine (and should be the preferred spelling) so that we provide a path forward for sha256 hashes in the API (and others).

>> Behavior change in warehouse implementation:
>> 
>> - if a project/release has no .info.requires_dist values, an empty list is returned. this is a change in that previously no key named info.requires_dist existed.
>
> I think the Warehouse behavior makes a lot more sense, and I highly suspect very little software out there is relying on the difference (and really, you shouldn't be relying on a key not existing in JSON output).

No disagreement from me in practice, but there is a _semantic_ difference between at least a `null` rather than an empty list.

**Comment by ewdurbin** [comment]
>> - .releases[][].path
>
> I don't see a path key? It shouldn't be exposed in either case, it's an implementation detail.

Whoops! You're right! it was actually removed. and that's fine :)

**Comment by dstufft** [comment]
These should be resolved now with #3213
